feat: flexible length specifications

RonMcKay opened the pull request:
This PR implements more flexible lengths for periods and step sizes. With this, meta tags like the following are possible:

`[x] [period] [@ YYYY-MM[-DD]] [/ [m] [step]]`

See also the updated README.

Additionally I fixed an error when the start date has no day assigned. (1310bc7) And added some tests for the new feature. (9469c83)
To make the specification more fluid in terms of the language I also added the possibility to specify `days`, `weeks`, `months` and `years`.

Note for @Akuukis: After this PR I will have a look into the Bug #18.

RonMcKay committed:
Fix error for begin dates without day

RonMcKay committed:
Add possibility for more flexible lengths

RonMcKay committed:
Add some tests for multipliers

RonMcKay committed:
Update README

RonMcKay commented:
> This feature is good!
> 
> Sadly it has a bug, split produces too much transactions when used as a tag, see the new tests pushed to master.

I found the bug. When the mark contained only an integer the regex matched that to be the multiplier and used the default period (`month`) resulting in too many txn. I updated the regex with two negative look aheads for the end of the string and for an '/'. Now a single integer is matched as the `period` and the multiplier is by default set to 1. I rebased everything on master and the tests are passing. Can you very the fix on your ledger?

Akuukis committed:
test(split): split few pennies over long time

RonMcKay commented:
Okay that was not related to #18 but the test case helped anyways. I forgot to add a positive lookahead for 'a digit boundary' in the regex. When someone uses the `x` together with `period` there has to always be some space or a letter following `x`. Otherwise the number needs to be matched as the period itself.
This caused in the test case the `17` to be split between `x` and `period`, resulting in `x=1` and `period=7`.
This should be fixed now, thanks for investigating!

Akuukis commented on beancount_interpolate/common.py:
```diff
@@ -118,7 +123,7 @@ def parse_mark(mark, default_date, config):
 
 # Infer Duration, start and steps. Spacing optinonal. Format: [123|KEYWORD] [@ YYYY-MM[-DD]] [/ step]
```
please update the comment here

Akuukis commented on beancount_interpolate/common.py:
```diff
@@ -118,7 +123,7 @@ def parse_mark(mark, default_date, config):
 # Infer Duration, start and steps. Spacing optinonal. Format: [123|KEYWORD] [@ YYYY-MM[-DD]] [/ step]
 # 0. max duration, 1. year, 2. month, 3. day, 4. min step
-RE_PARSING = re.compile(r"^\s*?([^-/\s]+)?\s*?(?:@\s*?([0-9]{4})-([0-9]{2})(?:-([0-9]{2}))?)?\s*?(?:\/\s*?([^-/\s]+)?\s*?)?$")
+RE_PARSING = re.compile(r"^\s*([0-9]+(?=[\sa-zA-Z]+)(?!\s*[@$/]))?\s*?([^-/\s]+)?\s*?(?:@\s*?([0-9]{4})-([0-9]{2})(?:-([0-9]{2}))?)?\s*?(?:\/\s*?([0-9]+)?\s*([^-/\s]+)?\s*?)?$")
```
This regex was awful to start with, but now it's beyond comprehensive. Can you please document it? For example like this 
```
E_PARSING = re.compile(r"^\s*([0-9]+(?=[\sa-zA-Z]+)(?!\s*[@$/]))?\s*?([^-/\s]+)?\s*?(?:@\s*?([0-9]{4})-([0-9]{2})(?:-([0-9]{2}))?)?\s*?(?:\/\s*?([0-9]+)?\s*([^-/\s]+)?\s*?)?$")
#                            ( 1.                              )     ( 2.     )
# 1. the first thing doing it's thing
# 2. the second thing matching another thing
# 3. ...
```

I'd also like to take it to regex-specific testing, and am curious about white-space handling between the parts as well.

RonMcKay replied:
Yes I can do that. I would then like to split the regex into multiple lines surrounded by braces.

RonMcKay replied:
This is then way easier to annotate

RonMcKay replied:
I also noticed that are some unnecessary `?` and maybe some other stuff.

RonMcKay replied:
A small test suite for different meta tags would be indeed quite nice.

Akuukis commented on beancount_interpolate/common.py:
```diff
@@ -143,6 +148,7 @@ def distribute_over_period(params, default_date, total_value, config):
     amounts = []
     accumulated_remainder = D(str(0))
 
+    # TODO: In the following `begin_date + i * step` can result in an OverflowError
```
please elaborate. Do I understand it correctly?
> in case `begin_date + duration` is so close to overflow, that last `begin_date + i * step` which would be beyond it and thus break the while loop raises OverflowError instead.

RonMcKay replied:
Yes, this would be a corner case. But theoretically one could also specify a step size of `inf` or something like `1 year / 10000 years` which would lead to an OverflowError. The second case is probably not that common so I think the bigger problem will be the more 'hidden' `OverflowError` when 
`begin_date + i * step < end_date` and `begin_date + i * step <= today_date`
but
`begin_date + (i + 1) * step > MAX_DATE`
This would then lead to an `OverflowError` in 
`tmp_date = begin_date + i * step`

This is however not a problem of the new feature as before there was also the possibility of overflowing the `datetime` objects.

RonMcKay replied:
I will document the TODO in the source code a little more in detail for future reference.

Akuukis commented:
Thanks you, now PR runs on my ledger without errors and with the same performance as before :heavy_check_mark: 

Thanks for pointing out the README, I rebased and force pushed to correct that. :heavy_check_mark: 

See few comments inline mostly about documentation. :spiral_notepad:

RonMcKay committed:
Update code documentation

Improved the documentation for `RE_PARSING` by splitting it into
multiple lines with description.
Additionally there was additional information added to the TODO item
describing a possible `OverflowError` in `distribute_over_period`.

RonMcKay commented:
@Akuukis I updated the code documentation and tried to add some description of the regex expression. I also added some further comments on the possible `OverflowError`.

RonMcKay committed:
Remove unnecessary `?` in `RE_PARSING`

Akuukis commented:
Thanks!
